### PR TITLE
Multi permit

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/udplib/udpsha.cpp [hunk 1]
```diff
@@ -34,18 +34,38 @@ using roxiemem::IDataBufferManager;
...
-bool udpDropDataPackets = false;
-unsigned udpDropFlowPackets[flowType::max_flow_cmd] = {};
-unsigned flowPacketsSent[flowType::max_flow_cmd] = {};
-#endif
+unsigned udpOutQsPriority = 0;
+unsigned udpMaxRetryTimedoutReqs = 0;   // 0 means off (keep retrying forever)
+
+//The following control the timeouts within the udp layer.  All timings are in milliseconds, but I suspect some of these should possibly be sub-millisecond
+unsigned udpFlowAckTimeout = 2;         // the maximum time that it is expected to take to receive an acknowledgement of a flow message (when one is sent) - should be small
+unsigned updDataSendTimeout = 20;       // how long to receive the maximum amount of data, ~100 packets of 8K should take 10ms on a 1Gb network. Timeout for assuming send_complete has been lost
```
Comment (richardkchapman) [inline roxie/udplib/udpsha.cpp:57]: Note: The comment suggests that this timeout needs to be dependent on the queue size (or the max slots per client) and possibly the MTU size in use.

File: roxie/udplib/udpsha.cpp [hunk 2]
```diff
@@ -34,18 +34,38 @@ using roxiemem::IDataBufferManager;
...
+unsigned udpRequestTimeout = 20;        // How long from a request for a permit until the permit is granted - timeout for assuming a ok_to_send has been lost.
+unsigned udpPermitTimeout = 50;         // How long is a grant expected to last before it is assumed lost?
+unsigned udpResendTimeout = 0;          // How long should elapse after a packet has been sent before we assume it is lost.
+                                        // 0 means they are unlikely to be lost, so worth resending as soon as it appears to be missing - trading duplicate packets for delays
+
+unsigned udpMaxPendingPermits = 1;      // default to 5?
+bool udpResendAllMissingPackets = false;  // Should probably default to true
```
Comment (richardkchapman) [inline roxie/udplib/udpsha.cpp:64]: Dubious about what this means, whether it defeats the overlapping of multiple bursts from a single sender, and what impact of latency on throughput might be if we put request_to_send_more on the data socket.

File: roxie/udplib/udpsha.cpp [hunk 3]
```diff
@@ -34,18 +34,38 @@ using roxiemem::IDataBufferManager;
...
+bool udpResendLostPackets = true;
+bool udpAssumeSequential = false;
+bool udpAdjustThreadPriorities = false;
```
Comment (richardkchapman) [inline roxie/udplib/udpsha.cpp:67]: This is a significant change to current behaviour...

File: roxie/udplib/udpsha.cpp [hunk 4]
```diff
@@ -293,6 +313,9 @@ int check_max_socket_write_buffer(int size) {
 #if defined( __linux__) || defined(__APPLE__)
 void setLinuxThreadPriority(int level)
 {
+    if (!udpAdjustThreadPriorities)
```
Comment (richardkchapman) [inline roxie/udplib/udpsha.cpp:316]: What about windows?

File: roxie/udplib/udpsha.hpp [hunk 5]
```diff
@@ -202,14 +202,29 @@ class simple_queue
...
+    enum flowCmd : unsigned short
+    {
+    //The first items in the flow command are also used as states in the receiver code:
+        send_completed,         // no activity from the sender - wating for a request to send
+        request_to_send,        // permit to be send has been requested but not granted
+        ok_to_send,             // permit has been granted
+
+    // The following messages
+        request_received,       // acknowledge request to send from the sender
+        send_start,             // about to send data - indicate how many packets are actually going to be send
```
Comment (richardkchapman) [inline roxie/udplib/udpsha.hpp:214]: typo: going to be send

File: roxie/udplib/udpsha.hpp [hunk 6]
```diff
@@ -202,14 +202,29 @@ class simple_queue
...
+        request_to_send_more,   // equivalent to send_completed followed by a request_to_send
```
Comment (richardkchapman) [inline roxie/udplib/udpsha.hpp:215]: May want to uncombine these if one should go on data socket (so as not to overtake associated data) but other should not.

File: roxie/udplib/udpsim.cpp [hunk 7]
```diff
@@ -132,16 +137,23 @@ void initOptions(int argc, const char **argv)
     udpDropFlowPackets[flowType::request_received] = options->getPropInt("@dropRequestReceivedPackets", 0);  // drop 1 in N
     udpDropFlowPackets[flowType::request_to_send] = options->getPropInt("@dropRequestToSendPackets", 0);  // drop 1 in N
     udpDropFlowPackets[flowType::request_to_send_more] = options->getPropInt("@dropRequestToSendMorePackets", 0);  // drop 1 in N
-    udpDropFlowPackets[flowType::send_completed] = options->getPropInt("@dropSendCompletedPackets", 0);  // drop 1 in N
+    udpDropFlowPackets[flowType::send_start] = options->getPropInt("@dropSendStartPackets", 0);  // drop 1 in N
+    udpDropFlowPackets[flowType::send_completed] = options->getPropInt("@   ", 0);  // drop 1 in N
```
Comment (richardkchapman) [inline roxie/udplib/udpsim.cpp:141]: bug: code looks corrupt here?

File: roxie/udplib/udpsim.cpp [hunk 8]
```diff
@@ -150,9 +162,10 @@ void initOptions(int argc, const char **argv)
 
 void simulateTraffic()
 {
-    constexpr unsigned numReceiveSlots = 100;
-    constexpr unsigned maxSlotsPerClient = 100;
-    constexpr unsigned maxSendQueueSize = 100;
+    //If a single permit allowed then grant that client all permits, otherwise allocate a share of twice the potential number of slots
+    const unsigned numReceiveSlots = 100;
+    const unsigned maxSlotsPerClient = (udpMaxPendingPermits == 1) ? numReceiveSlots : (2 * numReceiveSlots) / udpMaxPendingPermits;
```
Comment (richardkchapman) [inline roxie/udplib/udpsim.cpp:167]: Should this calculation be in send manager?

File: roxie/udplib/udpsim.cpp [hunk 9]
```diff
@@ -164,7 +177,7 @@ void simulateTraffic()
             unsigned header = 0;
             IpAddress pretendIP(VStringBuffer("8.8.8.%d", i));
             // Note - this is assuming we send flow on the data port (that option defaults true in roxie too)
-            Owned<ISendManager> sm = createSendManager(CCD_DATA_PORT, CCD_DATA_PORT, CCD_CLIENT_FLOW_PORT, maxSendQueueSize, 3, pretendIP, nullptr, false);
+            Owned<ISendManager> sm = createSendManager(CCD_SERVER_FLOW_PORT, CCD_DATA_PORT, CCD_CLIENT_FLOW_PORT, maxSendQueueSize, 3, pretendIP, nullptr, false);
```
Comment (richardkchapman) [inline roxie/udplib/udpsim.cpp:180]: Should probably be controlled by an option

File: roxie/udplib/udptrr.cpp [hunk 10]
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+            if (udpTraceLevel > 5 || udpTraceFlow)
+            {
+                StringBuffer ipStr;
+                DBGLOG("UdpReceiver: received %s msg flowSeq %" SEQF "u sendSeq %" SEQF "u from node=%s", flowType::name(msg.cmd), msg.flowSeq, msg.sendSeq, msg.sourceNode.getTraceText(ipStr).str());
+            }
+            UdpSenderEntry *sender = &parent.sendersTable[msg.sourceNode];
+
+            //If the flowSeq is lower than the previous flowSeq there could be two causes:
+            //a) The sender has restarted
+            //b) Messages have been received out of order (e.g. equest_to_send_more after a request_to_send?)
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:689]: typo: request

File: roxie/udplib/udptrr.cpp [hunk 11]
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+            //=> ignore the flow if it was the previous seq, and could not come from the server restarting
+            unsigned flowSeq = msg.flowSeq;
+            const unsigned startupFlowSeqs = 2; // flow seq below this could be from a restarting server
+            if ((flowSeq + 1 == sender->flowSeq) && (flowSeq >= startupFlowSeqs))
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:693]: Can you never be more than one ahead?

File: roxie/udplib/udptrr.cpp [hunk 12]
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+                            if (udpTraceLevel || udpTraceFlow || udpTraceTimeouts)
+                            {
+                                StringBuffer s;
+                                DBGLOG("permit to send %" SEQF "u to node %s timed out %u times - abandoning", finger->flowSeq, finger->dest.getIpText(s).str(), finger->timeouts);
+                            }
+                        }
+                        else
+                        {
+                            // Put it back on the queue (at the back)
+                            finger->permitTime = 0;
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:754]: bug: I don't understand why 0 (rather than now) would be wanted here

File: roxie/udplib/udptrr.cpp [hunk 13]
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+                    {
+                        timeout = udpPermitTimeout - elapsed;
+                        break;
+                    }
+                }
+
+                // Sum any remaining permits
+                while (finger)
+                {
+                    permitsIssued += finger->getActivePermits();
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:770]: Shouldn't we have accumulated these inside the loop above as well? I don't see where we do so.

**Review by richardkchapman — CHANGES_REQUESTED** [review]
First scan completed, need to look more later

File: roxie/udplib/udptrr.cpp (hunk 12 above)
Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:754]: Ok, then perhaps would be clearer/better to move this line to just after pendingPermits.remove(finger) ?

Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:754]: (and set finger->activePermits to 0)

File: roxie/udplib/udptrr.cpp (hunk 13 above)
Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:770]: ok

File: roxie/udplib/udptrr.cpp (hunk 11 above)
Comment (ghalliday) [reply roxie/udplib/udptrr.cpp:693]: I think so, but needs to be checked, and rethought if completed flow messages are sent on the flow thread.

File: roxie/udplib/udptrr.cpp [hunk 14]
```diff
@@ -171,23 +610,24 @@ class CReceiveManager : implements IReceiveManager, public CInterface
                 if (udpAssumeSequential && packetsSeen.hasGaps())
                     return true;
             }
-            if (msTick()-requestTime > udpResendTimeout)
-                return true;
-            return false;
+
+            //The best approximation to the oldest-inlight packet - because permits may have expired...
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:614]: typo: inlight

File: roxie/udplib/udptrr.cpp [hunk 15]
```diff
@@ -94,6 +492,43 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+            flowSeq = 0;
+            sendSeq = 0;
+            permitTime = 0;
+            numPackets.store(0, std::memory_order_release);
+        }
+    };
+
+    //Increasing this number, increases the number of concurrent permits a sender may have (and its resilience to lost flow messages),
+    //but also increases the processing cost since code often iterates through all the permits.  2..4 likely to be good values.
+    static constexpr unsigned MAX_PERMITS = 2;
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:530]: Might be better to call it MAX_PERMITS_PER_SENDER

File: roxie/udplib/udptrr.cpp [hunk 16]
```diff
@@ -270,71 +710,118 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+            }
+            return 0;
+        }
 
-    class SenderList
-    {
-        UdpSenderEntry *head = nullptr;
-        UdpSenderEntry *tail = nullptr;
-        unsigned numEntries = 0;
+        //Total permits outstanding for the sender
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:727]: Total reservations, I think. There is scope for confusion when using the term 'permit' between an active permission to send and the number of slots that the permit grants...

File: roxie/udplib/udptrr.cpp [hunk 17]
```diff
@@ -270,71 +710,118 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+                if (permits[i].flowSeq == flowSeq)
+                {
+                    permits[i].numPackets.store(num, std::memory_order_release);
+                    return;
+                }
+            }
+        }
+        inline void decPermit(unsigned msgSeq)
+        {
+            //Although this is a bit more work than matching by flowSeq it shouldn't be too inefficient.
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:797]: Usually the permit will be the same one that we matched last time. Can we do a quick check to tell?

Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:797]: This code feels painful to be executing per received message

File: roxie/udplib/udptrr.cpp [hunk 18]
```diff
@@ -433,20 +971,195 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+                // Sum the number of permits that are allocated to active senders
+                while (finger)
+                {
+                    permitsIssued += finger->getNumReserved();
+                    finger = finger->next;
+                }
+            }
+
+            // Aim is to issue enough permits to use all available the space in the queue.
+            // As data is received the number of slots reduces, but the relevant permit is reduced to match
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:1080]: I don't really understand this comment

File: roxie/udplib/udptrr.cpp [hunk 19]
```diff
@@ -433,20 +971,195 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+            if (slots >= permitsIssued)
+                slots -= permitsIssued;
+            else
+                slots = 0;
+
+            bool anyCanSend = false;
+            bool anyCannotSend = false;
+            for (UdpSenderEntry *finger = pendingRequests; finger != nullptr; finger = finger->next)
+            {
+                //MORE: not sure if the permitsIssued test is required
```
Comment (richardkchapman) [inline roxie/udplib/udptrr.cpp:1091]: MORE: should work it out then!

**Review by richardkchapman — COMMENTED** [review]
General comment: this is beyond the scope of 8.4.x point release - but there may be some changes within it that need to go in 8.4 ?

**Comment by ghalliday** [comment]
@richardkchapman Closing to open a new PR without all the cleaned up comments.  Probably helpful if you read through this PR to check I have addressed all the comments though.

**Pull request closed without merging**